src/librustc/infer/mod.rs

```diff
@@ -1447,14 +1447,23 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                                    expected: Ty<'tcx>,
                                    actual: Ty<'tcx>,
                                    err: TypeError<'tcx>) {
+        self.mismatched_types_diag(origin, expected, actual, err).emit();
+    }
+
+    pub fn mismatched_types_diag(&self,
+                                 origin: TypeOrigin,
+                                 expected: Ty<'tcx>,
+                                 actual: Ty<'tcx>,
+                                 err: TypeError<'tcx>)
+                                 -> DiagnosticBuilder<'tcx> {
         let trace = TypeTrace {
             origin: origin,
             values: Types(ExpectedFound {
                 expected: expected,
                 found: actual
             })
         };
-        self.report_and_explain_type_error(trace, &err).emit();
+        self.report_and_explain_type_error(trace, &err)
     }
```

Comment:
If you added this function you should use it wherever `report_and_explain_type_error` was used for the same effect.